File: mvp-java/src/main/java/com/wf/mvp/java/view/NFragment.java
package com.wf.mvp.java.view;

import android.annotation.SuppressLint;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Message;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Toast;

import androidx.fragment.app.Fragment;

import com.wf.mvp.java.customize.UiHandler;
import com.wf.mvp.java.impl.Impl;

/**
 * MvpMode -> com.wf.mvp.java.view -> NFragment
 *
 * @Author: wf-pc
 * @Date: 2020-05-27 18:19
 * <p>
 * Mvp's v-layer basic no presenter Fragment, used to implement common functions.
 */
abstract public class NFragment extends Fragment implements Impl {

    /**
     * The instance object of UI thread handler.
     */
    protected UiHandler mUiHandler;

    private Toast mToast;
    private ProgressDialog mLoading;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        beforeCreate();
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(bindLayoutId(), container, false);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mUiHandler = new UiHandler();
        mUiHandler.attachRefs(this);
        Intent intent = getActivity() == null ? new Intent() : getActivity().getIntent();
        initView(view, intent);
        initListener();
        initData();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mUiHandler != null) {
            mUiHandler.detachRefs();
            mUiHandler = null;
        }
        release();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }


    /**
     * Used to do something before Create.
     */
    protected void beforeCreate() {}

    /**
     * Used to bind layout files.
     */
    protected abstract int bindLayoutId();

    /**
     * Used to initialize view.
     */
    protected abstract void initView(View layout, Intent intent);

    /**
     * Used to initialize listener.
     */
    protected abstract void initListener();

    /**
     * Used to initialize data.
     */
    protected abstract void initData();

    /**
     * Used to release resources.
     */
    protected abstract void release();

    /**
     * Used to handle message.
     */
    @Override
    abstract public void onHandleMessage(Message msg);


    @Override
    public void runOnUiThread(Runnable runnable) {
        if (getActivity() == null) return;
        getActivity().runOnUiThread(runnable);
    }

    @Override
    public int getColor(int resId) {
        if (getActivity() == null) return 0;
        return getActivity().getResources().getColor(resId);
    }

    @Override
    public float getDimension(int resId) {
        if (getActivity() == null) return 0;
        return getActivity().getResources().getDimension(resId);
    }

    @Override
    public Drawable getDrawable(int resId) {
        if (getActivity() == null) return null;
        return getActivity().getResources().getDrawable(resId);
    }


    @SuppressLint("ShowToast")
    @Override
    public void showToast(String message) {
        if (getActivity() == null) return;
        if (mToast == null) {
            mToast = Toast.makeText(getActivity(), "", Toast.LENGTH_SHORT);
        }
        mToast.setText(message);
        mToast.show();
    }

    @SuppressLint("ShowToast")
    @Override
    public void showLongToast(String message) {
        if (getActivity() == null) return;
        if (mToast == null) {
            mToast = Toast.makeText(getActivity(), "", Toast.LENGTH_LONG);
        }
        mToast.setText(message);
        mToast.show();
    }

    @Override
    public void showLoading() {
        showLoading("");
    }

    @Override
    public void showLoading(String message) {
        if (getActivity() == null) return;
        mLoading = ProgressDialog.show(getActivity(), "", message);
    }

    @Override
    public void hideLoading() {
        if (mLoading != null && mLoading.isShowing()) {
            mLoading.dismiss();
        }
    }

    @Override
    public void toggleKeyboard() {
        if (getActivity() == null) return;
        InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null && getActivity().getCurrentFocus() != null
                && getActivity().getCurrentFocus().getWindowToken() != null) {
            imm.toggleSoftInputFromWindow(
                    getActivity().getCurrentFocus().getWindowToken(),
                    0, InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }

    @Override
    public void showKeyboard(View view) {
        if (getActivity() == null || view == null) return;
        view.requestFocus();
        final InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.showSoftInput(view, InputMethodManager.SHOW_FORCED);
        }
    }

    @Override
    public void hideKeyboard() {
        if (getActivity() == null) return;
        InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null && imm.isActive() && getActivity().getCurrentFocus() != null
                && getActivity().getCurrentFocus().getWindowToken() != null) {
            imm.hideSoftInputFromWindow(getActivity().getCurrentFocus().getWindowToken(),
                    InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }
}

File: mvp-kotlin/src/main/java/com/wf/mvp/kotlin/view/NFragment.kt
package com.wf.mvp.kotlin.view

import android.annotation.SuppressLint
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Message
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.wf.mvp.kotlin.customize.UiHandler
import com.wf.mvp.kotlin.impl.Impl

/**
 * MvpMode -> com.wf.mvp.kotlin.view -> AbsFragment
 * @Author: wf-pc
 * @Date: 2020-05-18 21:59
 * <p>
 * Mvp's v-layer basic no presenter Fragment, used to implement common functions.
 */
abstract class NFragment : Fragment(), Impl {

    /**
     * The instance object of UI thread handler.
     */
    protected var mUiHandler: UiHandler? = null

    private var mToast: Toast? = null
    private var mLoading: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        beforeCreate()
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(bindLayoutId(), container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mUiHandler = UiHandler()
        mUiHandler?.attachRefs(this)
        val intent = activity?.intent ?: Intent()
        initView(view, intent)
        initListener()
        initData()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mUiHandler?.detachRefs()
        mUiHandler = null
        release()
    }

    override fun onDestroy() {
        super.onDestroy()
    }


    /**
     * Used to do something before Create.
     */
    protected open fun beforeCreate() {}

    /**
     * Used to bind layout files.
     */
    protected abstract fun bindLayoutId(): Int

    /**
     * Used to initialize view.
     */
    protected abstract fun initView(layout: View, intent: Intent);

    /**
     * Used to initialize listener.
     */
    protected abstract fun initListener();

    /**
     * Used to initialize data.
     */
    protected abstract fun initData();

    /**
     * Used to release resources.
     */
    protected abstract fun release();

    /**
     * Used to handle message.
     */
    abstract override fun onHandleMessage(msg: Message)


    override fun runOnUiThread(block: () -> Unit) {
        mUiHandler?.post(block)
    }

    override fun getDimension(resId: Int): Float {
        return resources.getDimension(resId)
    }

    @SuppressLint("ShowToast")
    override fun showToast(message: String) {
        if (activity != null) {
            if (mToast == null) {
                mToast = Toast.makeText(activity, "", Toast.LENGTH_SHORT)
            }
            mToast?.setText(message)
            mToast?.show()
        }
    }

    @SuppressLint("ShowToast")
    override fun showLongToast(message: String) {
        if (activity != null) {
            if (mToast == null) {
                mToast = Toast.makeText(activity, "", Toast.LENGTH_LONG)
            }
            mToast?.setText(message)
            mToast?.show()
        }
    }

    override fun showLoading() {
        showLoading("")
    }

    override fun showLoading(message: String) {
        if (activity != null) {
            mLoading = ProgressDialog.show(activity, "", message)
        }
    }

    override fun hideLoading() {
        mLoading?.dismiss()
    }

    override fun toggleKeyboard() {
        val imm = activity?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?
        imm?.toggleSoftInputFromWindow(
            activity?.currentFocus?.windowToken,
            0,
            InputMethodManager.HIDE_NOT_ALWAYS
        )
    }

    override fun showKeyboard(view: View) {
        view.requestFocus()
        val imm = activity?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?
        imm?.showSoftInput(view, InputMethodManager.SHOW_FORCED)
    }

    override fun hideKeyboard() {
        val imm = activity?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?
        imm?.hideSoftInputFromWindow(
            activity?.currentFocus?.windowToken,
            InputMethodManager.HIDE_NOT_ALWAYS
        )
    }
}